src/v/kafka/server/handlers/produce.cc

```diff
@@ -214,55 +236,104 @@ static ss::future<produce_response::partition> produce_topic_partition(
-            acks,
-            num_records);
-      });
-    return f.then([&octx, start](produce_response::partition p) {
-        if (p.error_code == error_code::none) {
-            auto dur = std::chrono::steady_clock::now() - start;
-            octx.rctx.connection()->server().update_produce_latency(dur);
-        }
-        return p;
-    });
+
+    auto dispatch = ss::make_foreign<std::unique_ptr<ss::promise<>>>(
+      std::make_unique<ss::promise<>>());
+    auto dispatch_f = dispatch->get_future();
+    auto f
+      = octx.rctx.partition_manager()
+          .invoke_on(
+            *shard,
+            octx.ssg,
+            [reader = std::move(reader),
+             ntp = std::move(ntp),
+             dispatch = std::move(dispatch),
+             num_records,
+             bid,
+             acks = octx.request.data.acks,
+             source_shard = ss::this_shard_id()](
+              cluster::partition_manager& mgr) mutable {
+                auto partition = mgr.get(ntp);
+                if (!partition) {
+                    // submit back to promise source shard
+                    return ss::smp::submit_to(
+                      source_shard,
+                      [dispatch = std::move(dispatch),
+                       pid = ntp.tp.partition]() mutable {
+                          dispatch->set_value();
+                          return produce_response::partition{
+                            .partition_index = pid,
+                            .error_code
+                            = error_code::unknown_topic_or_partition};
+                      });
+                }
+                if (unlikely(!partition->is_leader())) {
+                    // submit back to promise source shard
+                    return ss::smp::submit_to(
+                      source_shard,
+                      [dispatch = std::move(dispatch),
+                       pid = ntp.tp.partition]() mutable {
+                          dispatch->set_value();
+                          return produce_response::partition{
+                            .partition_index = pid,
+                            .error_code = error_code::not_leader_for_partition};
+                      });
+                }
+                auto stages = partition_append(
+                  ntp.tp.partition,
+                  ss::make_lw_shared<replicated_partition>(
+                    std::move(partition)),
+                  bid,
+                  std::move(reader),
+                  acks,
+                  num_records);
+                return stages.dispatched
+                  .then_wrapped([source_shard, dispatch = std::move(dispatch)](
+                                  ss::future<> f) mutable {
+                      if (f.failed()) {
+                          return ss::smp::submit_to(
+                            source_shard,
+                            [dispatch = std::move(dispatch),
+                             e = f.get_exception()]() mutable {
```

Comment:
this is probably safe, but I dunno. did you take a look at `get_exception`? there are a couple versions, and one is `const std::exception_ptr& get_exception()` and I'm really not sure if this is transmitting a reference to any of the bits in `f` back across the core in this lambda capture.

Reply:
The only implementation I see is:
```c++
  [[gnu::always_inline]]
    std::exception_ptr get_exception() noexcept {
        return get_available_state_ref().get_exception();
    }
```
It returns a copy, it is safe to access this from other core.